libpod/container.go

```diff
@@ -107,10 +108,19 @@ type Container struct {
 	batched bool
 
 	valid      bool
-	lock       lock.Locker
 	runtime    *Runtime
 	ociRuntime OCIRuntime
 
+	// This is a lock (usually in-memory, but may be a file lock) providing
+	// multi-process mutual exclusive for containers. All operations working
+	// with container state should take and hold this lock.
+	lock lock.Locker
+	// This is a single-process lock, used to try and prevent multiple
+	// threads of the same Podman process from removing the same container
+	// at the same time (when using recursive removal of dependencies).
+	// This probably should not be used outside of `removeContainer`.
+	perProcessLock sync.Mutex
```

Comment:
To prevent ABBA deadlocks properly worth to clarify that this lock should only be taken after the normal container lock.